product/util.py

```diff
@@ -171,164 +211,197 @@ def diff_products(left_product, right_product, ignore_component_names=()):
-    names_version_changed: set = dataclasses.field(default_factory=set)
-
-
-def diff_images(
-    left_component_descriptor,
-    right_component_descriptor,
-    left_component,
-    right_component
-):
-    left_images = set(_effective_images(left_component_descriptor, left_component))
-    right_images = set(_effective_images(right_component_descriptor, right_component))
+class ResourceDiff:
+    left_component: cm.Component
+    right_component: cm.Component
+    resource_refs_only_left: typing.List[cm.Resource] = dataclasses.field(default_factory=list)
+    resource_refs_only_right: typing.List[cm.Resource] = dataclasses.field(default_factory=list)
+    resourcepairs_version_changed: typing.List[typing.Tuple[cm.Resource, cm.Resource]] = dataclasses.field(default_factory=list) # noqa:E501
+
+
+def _add_if_not_duplicate(list, res):
+    if res.name not in list:
+        list.append(res)
+
+
+def diff_resources(
+    left_component: cm.Component,
+    right_component: cm.Component,
+) -> ResourceDiff:
+    if type(left_component) is not cm.Component:
+        raise NotImplementedError(
+            f'unsupported {type(left_component)=}. Only cm.Component is supported',
+        )
+    if type(right_component) is not cm.Component:
+        raise NotImplementedError(
+            f'unsupported {type(right_component)=}. Only cm.Component is supported',
+        )
 
-    left_names_to_imgs = {i.name(): i for i in left_images}
-    right_names_to_imgs = {i.name(): i for i in right_images}
+    left_resource_identities = {
+        r.identity(left_component.resources + right_component.resources): r
+        for r in left_component.resources
+    }
+    right_resource_identities_to_resource = {
+        r.identity(left_component.resources + right_component.resources): r
+        for r in right_component.resources
+    }
 
-    img_diff = ImageDiff(
+    resource_diff = ResourceDiff(
         left_component=left_component,
         right_component=right_component,
     )
 
-    if left_images == right_images:
-        return img_diff
-
-    for name, img in left_names_to_imgs.items():
-        if not name in right_names_to_imgs:
-            img_diff.irefs_only_left.add(img)
+    if left_resource_identities.keys() == right_resource_identities_to_resource.keys():
+        return resource_diff
+
+    left_names_to_resource = {r.name: r for r in left_component.resources}
+    right_names_to_resource = {r.name: r for r in right_component.resources}
+    # get left exclisive resources
+    for resource in left_resource_identities.values():
+        if not resource.name in right_names_to_resource:
+            _add_if_not_duplicate(resource_diff.resource_refs_only_left, resource)
+
+    # get right exclusive images
+    for resource in right_resource_identities_to_resource.values():
+        if not resource.name in left_names_to_resource:
+            _add_if_not_duplicate(resource_diff.resource_refs_only_right, resource)
+
+    def enumerate_group_pairs(
```

Comment:
by now, i really feel bad about having implemented this with so few comments. e.g. a brief comment explaining why we build resource-group-pairs might help, wdyt (not help you, any more, since you already figured out the algorithm, obviously)?